pace-util/pace/util/report.py

```diff
@@ -0,0 +1,114 @@
+    return experiment
+
+
+def collect_keys_from_data(times_per_step: List[Dict[str, float]]) -> List[str]:
+    """Collects all the keys in the list of dics and returns a sorted version"""
+    keys = set()
+    for data_point in times_per_step:
+        for k, _ in data_point.items():
+            keys.add(k)
+    sorted_keys = list(keys)
+    sorted_keys.sort()
+    return sorted_keys
+
+
+def gather_timing_data(
+    times_per_step: List[Dict[str, float]],
+    results: Dict[str, Any],
+    comm: MPI.Comm,
+    root: int = 0,
+) -> Dict[str, Any]:
+    """returns an updated version of  the results dictionary owned
+    by the root node to hold data on the substeps as well as the main loop timers"""
+    is_root = comm.Get_rank() == root
+    keys = collect_keys_from_data(times_per_step)
+    data: List[float] = []
+    for timer_name in keys:
+        data.clear()
+        for data_point in times_per_step:
+            if timer_name in data_point:
+                data.append(data_point[timer_name])
+
+        sendbuf = np.array(data)
+        recvbuf = None
+        if is_root:
+            recvbuf = np.array([data] * comm.Get_size())
+        comm.Gather(sendbuf, recvbuf, root=0)
+        if is_root:
+            results["times"][timer_name]["times"] = copy.deepcopy(recvbuf.tolist())
+    return results
+
+
+def write_global_timings(experiment: Dict[str, Any]) -> None:
+    now = datetime.now()
+    filename = now.strftime("%Y-%m-%d-%H-%M-%S")
+    with open(filename + ".json", "w") as outfile:
+        json.dump(experiment, outfile, sort_keys=True, indent=4)
+
+
+def gather_hit_counts(
+    hits_per_step: List[Dict[str, int]], results: Dict[str, Any]
+) -> Dict[str, Any]:
+    """collects the hit count across all timers called in a program execution"""
+    for data_point in hits_per_step:
+        for name, value in data_point.items():
+            if name not in results["times"]:
+                print(name)
+                results["times"][name] = {"hits": value, "times": []}
+            else:
+                results["times"][name]["hits"] += value
+    return results
+
+
+def collect_data_and_write_to_file(
+    time_step: int,
+    backend: str,
+    git_hash: str,
+    comm: MPI.Comm,
+    hits_per_step: List,
+    times_per_step: List,
+    experiment_name: str,
+    dt_atmos: float,
+) -> None:
+    """
+    collect the gathered data from all the ranks onto rank 0 and write the timing file
+    """
+    is_root = comm.Get_rank() == 0
+    results = None
+    if is_root:
+        print("Gathering Times")
+        results = set_experiment_info(experiment_name, time_step, backend, git_hash)
+        results = gather_hit_counts(hits_per_step, results)
+
+    results = gather_timing_data(times_per_step, results, comm)
+
+    if is_root:
+        mainloop = np.mean(sum(results["times"]["mainloop"]["times"], []))
+        speedup = dt_atmos / mainloop
+        SYPD = 1.0 / 365.0 * speedup
+        results["SYPD"] = SYPD
```

Comment:
nit: Can you refactor this variable name to be lowercase? The syntax highlighting messes with me.